File: 2.트리/트리순회1991.cpp
// #include <vector>
// #include <algorithm>
// #include <iostream>
// using namespace std;
// struct Node
// {
//     char alp;
//     char left;
//     char right;
// };
// Node a[30];
// void preorder(char node,vector<char> &answer)
// {
//     if(node=='0')
//     {
//         return;
//     }
//     answer.push_back(a[node-'A'].alp);
//     preorder(a[node-'A'].left,answer);
//     preorder(a[node-'A'].right,answer);
// }
// void inorder(char node,vector<char> &answer)
// {
//     if(node=='0')
//     {
//         return;
//     }
//     inorder(a[node-'A'].left,answer);
//     answer.push_back(a[node-'A'].alp);
//     inorder(a[node-'A'].right,answer);
// }
// void postorder(char node,vector<char> &answer)
// {
//     if(node=='0')
//     {
//         return;
//     }
//     postorder(a[node-'A'].left,answer);
//     postorder(a[node-'A'].right,answer);
//     answer.push_back(a[node-'A'].alp);
// }
// int main()
// {
//     int n;
//     cin >> n;
//     for(int i=0;i<n;i++)
//     {
//         Node temp;
//         char x,y,z;
//         cin >> x >> y >> z;
//         temp.alp = x;
//         if(y=='.') temp.left = '0';
//         else temp.left = y;
//         if(z=='.') temp.right = '0';
//         else temp.right = z;
//         a[x-'A'] = temp;  
//     }
//     vector<char> answer[3];
//     preorder('A',answer[0]);
//     inorder('A',answer[1]);
//     postorder('A',answer[2]);
//     for(int i=0;i<3;i++)
//     {
//         for(int j=0;j<answer[i].size();j++)
//         {
//             cout << answer[i][j];
//         }
//         cout << '\n';
//     }
//     return 0;
// }
#include <bits/stdc++.h>
using namespace std;
struct Node {
  char left = NULL;
  char right = NULL;
};
Node a[30];

void preorder(char root, vector<char> &answer) {
  if(root==NULL) return;
  answer.push_back(root);
  preorder(a[root-'A'].left,answer);
  preorder(a[root-'A'].right,answer);
}

void inorder(char root, vector<char> &answer) {
  if(root==NULL) return;
  inorder(a[root-'A'].left,answer);
  answer.push_back(root);
  inorder(a[root-'A'].right,answer);
}

void postorder(char root, vector<char> &answer) {
  if(root==NULL) return;
  postorder(a[root-'A'].left,answer);
  postorder(a[root-'A'].right,answer);
  answer.push_back(root);
}

int main()
{
  int n;
  cin >> n;
  for(int i=0;i<n;i++) {
    char x,y,z;
    cin >> x >> y >> z;
    Node temp;
    if(y!='.') temp.left = y;
    if(z!='.') temp.right = z;
    a[x-'A'] = temp;
  }
  vector<char> answer[3];
  preorder('A',answer[0]);
  inorder('A',answer[1]);
  postorder('A',answer[2]);
  for(int i=0;i<3;i++) {
    for(int j=0;j<answer[i].size();j++) {
      cout << answer[i][j];
    }
    cout << '\n';
  }
  return 0;
}

File: Java/2.트리/트리순회1991.java
//import java.io.BufferedReader;
//import java.io.IOException;
//import java.io.InputStreamReader;
//import java.util.StringTokenizer;
//
//public class Main {
//
//	static class Node {
//		char left = '0';
//		char right = '0';
//	}
//
//	static void preorder(char root) {
//		if(root=='0') {
//			return;
//		}
//		System.out.print(root);
//		preorder(a[root-'A'].left);
//		preorder(a[root-'A'].right);
//	}
//
//	static void inorder(char root) {
//		if(root=='0') {
//			return;
//		}
//		inorder(a[root-'A'].left);
//		System.out.print(root);
//		inorder(a[root-'A'].right);
//	}
//
//	static void postorder(char root) {
//		if(root=='0') {
//			return;
//		}
//		postorder(a[root-'A'].left);
//		postorder(a[root-'A'].right);
//		System.out.print(root);
//	}
//
//	static Node[] a = new Node[30];
//
//	public static void main(String[] args) throws IOException {
//		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
//		StringBuilder sb = new StringBuilder();
//		StringTokenizer st;
//
//		int n = Integer.parseInt(br.readLine());
//		for(int i=0;i<n;i++) {
//			st = new StringTokenizer(br.readLine());
//			char x = st.nextToken().charAt(0);
//			char y = st.nextToken().charAt(0);
//			char z = st.nextToken().charAt(0);
//			Node temp = new Node();
//			if(y!='.') temp.left = y;
//			if(z!='.') temp.right = z;
//			a[x-'A'] = temp;
//		}
//		preorder('A');
//		System.out.println();
//		inorder('A');
//		System.out.println();
//		postorder('A');
//	}
//}
import java.io.*;
import java.util.*;

public class Main {
	
	static class Node {
		char v;
		char left;
		char right;
		Node(char v, char left, char right) {
			this.v = v;
			this.left = left;
			this.right = right;
		}
	}
	static Node[] a = new Node[27];
	
	static void preorder(char root) {
		if(root=='.') return;
		System.out.print(a[root-'A'].v);
		preorder(a[root-'A'].left);
		preorder(a[root-'A'].right);
	}
	
	static void inorder(char root) {
		if(root=='.') return;
		inorder(a[root-'A'].left);
		System.out.print(a[root-'A'].v);
		inorder(a[root-'A'].right);
	}
	
	static void postorder(char root) {
		if(root=='.') return;
		postorder(a[root-'A'].left);
		postorder(a[root-'A'].right);
		System.out.print(a[root-'A'].v);
	}
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st;
		
		int n = Integer.parseInt(br.readLine());
		for(int i=0;i<n;i++) {
			st = new StringTokenizer(br.readLine());
			char x = st.nextToken().charAt(0);
			char y = st.nextToken().charAt(0);
			char z = st.nextToken().charAt(0);
			Node temp = new Node(x,y,z);
			a[x-'A'] = temp;
		}
		preorder('A');
		System.out.println();
		inorder('A');
		System.out.println();
		postorder('A');
	}
}
